query-engine/js-connectors/src/proxy.rs

```diff
@@ -316,40 +282,26 @@ fn js_base_value_to_quaint(json_value: serde_json::Value, column_type: ColumnTyp
     }
 }
 
-// TODO: this flavour-specific deserialization logic needs to be moved to the JS part of the connector
-pub struct FlavourSpecificResultSet(pub (Flavour, JSResultSet));
-
-impl From<FlavourSpecificResultSet> for QuaintResultSet {
-    fn from(pair: FlavourSpecificResultSet) -> Self {
-        let (flavour, mut js_result_set) = pair.0;
-        // TODO: extract, todo: error rather than panic?
-        let to_quaint_row = move |row: &mut Vec<serde_json::Value>| -> Vec<quaint::Value<'static>> {
-            let mut res = Vec::with_capacity(row.len());
-
-            for i in 0..row.len() {
-                let column_type = js_result_set.column_types[i];
-                let json_value = row.remove(0);
-
-                // Note: here, we could consider using conditional compile-time variables to avoid the match.
-                let quaint_value = match flavour {
-                    Flavour::Mysql => js_planetscale_value_to_quaint(json_value, column_type),
-                    Flavour::Postgres => js_neon_value_to_quaint(json_value, column_type),
-                    _ => unreachable!(
-                        "Unsupported flavour {:?} in result set transformation from javascript",
-                        flavour
-                    ),
-                };
-
-                res.push(quaint_value);
-            }
-
-            res
-        };
-
-        let names = js_result_set.column_names;
-        let rows = js_result_set.rows.iter_mut().map(to_quaint_row).collect();
-
-        QuaintResultSet::new(names, rows)
+impl From<JSResultSet> for QuaintResultSet {
+    fn from(js_result_set: JSResultSet) -> Self {
+        let JSResultSet {
+            rows,
+            column_names,
+            column_types,
+        } = js_result_set;
+
+        let quaint_rows = rows
+            .into_iter()
+            .map(move |row| {
+                column_types
+                    .iter()
+                    .zip(row)
+                    .map(|(column_type, value)| js_value_to_quaint(value, *column_type))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
```

Comment:
Note: this logic was originally written in an imperative way (rather than the current, functional approach) by Miguel for performance reasons. Was that a micro-optimisation? Maybe, I haven't seen the most up to date benchmarks.

I find the functional style proposed by @SevInf indeed easier to visually parse, and we can always revisit this logic should we find out performance reasons to do so.

Reply:
From what I remember in Rust Book, iterators produce the code that is almost identical to manual iteration, but I indeed did not measure it.